### PR TITLE
feat: implement FollowUserAsync method in FriendService

### PR DESCRIPTION
**Pull request by jhael-arce-jalafund**
## **Trello Ticket**

[ID-14](https://trello.com/c/HDQRuXk9/19-id-14following-a-user)

---

## What did I do?

I implemented functionality for users to follow and unfollow other users, along with the ability to view their friends list. This enhancement facilitates user engagement by allowing them to curate their feeds based on their connections.

- Created REST endpoints in the `FriendsController` for following and unfollowing users, as well as retrieving a user's friends list.
- Developed the `FollowUserAsync` and `UnfollowUserAsync` methods in the `FriendService` to manage follow/unfollow operations.
- Implemented validation in the `FriendsRepository` to prevent users from following themselves.
- Ensured that successful follow and unfollow operations return appropriate success messages.
- Created `FriendResponse` to structure the friends list response, including user ID and name.

---

## Why did I do this?

This functionality was necessary to enhance user interaction and customization of their feeds.

- **Problem:** Users needed a way to manage who they follow, affecting the posts they see in their feed. Without this, users would see posts from all users, making the feed cluttered and less personalized.
- **Value added:** By enabling follow/unfollow actions and displaying a friends list, users can tailor their social experience, improving overall engagement and satisfaction with the platform.

---

## How did I do it?

- Defined new endpoints:
  - `POST /users/{userId}/follow` to follow a user.
  - `DELETE /users/{userId}/unfollow` to unfollow a user.
  - `GET /users/{userId}/friends` to retrieve the user's friends list.
  
- Implemented the `FollowUserAsync` and `UnfollowUserAsync` methods in `FriendService` for business logic related to following and unfollowing users.
- Enhanced the `FriendsRepository` with database operations to persist follow/unfollow actions and retrieve the friends list.
- Introduced validation to ensure users cannot follow themselves.
- Utilized `FriendResponse` to map and return the friends list, ensuring consistency in the output.

---

### CODE REVIEW

File: SocialNetworkApi/Controllers/CommentsController.cs [hunk 1]
```diff
@@ -1,55 +1,54 @@
...
-        }
+        _commentService = commentService;
+    }
 
+    [HttpPost]
+    public async Task<IActionResult> Create(Guid postId, [FromBody] CreateCommentRequest request)
+    {
+        var comment = request.ToDomain(postId);
+        var result = await _commentService.CreateAsync(comment);
+        return CreatedAtAction(nameof(GetAllComments), new { postId, commentId = result.Data.CommentId }, result.Data);
```
Comment (GabyGV-GGV) [inline SocialNetworkApi/Controllers/CommentsController.cs:29]: Dereference of a possibly null reference.

File: SocialNetworkApi/DataAccess/Entities/Comment.cs [hunk 2]
```diff
@@ -1,17 +1,17 @@
...
+    public Guid UserId { get; set; }
+    public string Content { get; set; }
+    public DateTime CreatedAt { get; set; }
 
-        public Comment()
-        {
-            Id = Guid.NewGuid();
-            CreatedAt = DateTime.UtcNow;
-        }
+    public Comment()
```
Comment (GabyGV-GGV) [inline SocialNetworkApi/DataAccess/Entities/Comment.cs:11]: Non-nullable property 'Content' must contain a non-null value when exiting constructor. Consider adding the 'required' modifier or declaring the property as nullable.

File: SocialNetworkApi/DataAccess/Repositories/Concretes/CommentRepository.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    public async Task<IEnumerable<Comment>> GetByPostIdAsync(Guid postId)
+    {
+        return await _dbContext.Comments
+            .Where(c => c.PostId == postId)
+            .ToListAsync();
+    }
+
+    public async Task<Comment> GetByIdAsync(Guid commentId) 
+    {
+        return await _dbContext.Comments.FindAsync(commentId);
```
Comment (GabyGV-GGV) [inline SocialNetworkApi/DataAccess/Repositories/Concretes/CommentRepository.cs]: Possible null reference return.

File: SocialNetworkApi/DataAccess/Repositories/Concretes/CommentRepository.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    public async Task<Comment> UpdateAsync(Guid commentId, Comment updatedComment)
+    {
+        var existingComment = await _dbContext.Comments.FindAsync(commentId);
+        if (existingComment == null)
+            return null;
```
Comment (GabyGV-GGV) [inline SocialNetworkApi/DataAccess/Repositories/Concretes/CommentRepository.cs]: Possible null reference return.

File: SocialNetworkApi/DataAccess/Repositories/Concretes/FriendsRepository.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            .Join(_dbContext.Users, friendId => friendId, user => user.UserId, (friendId, user) => user)
+            .ToListAsync();
+    }
+
+    public async Task<bool> AreFriendsAsync(Guid userId, Guid friendId)
+    {
+        return await _dbContext.Friends
+            .AnyAsync(f => f.UserId == userId && f.FriendId == friendId);
+    }
+}
```
Comment (GabyGV-GGV) [inline SocialNetworkApi/DataAccess/Repositories/Concretes/FriendsRepository.cs]: Do not delete the final line

File: SocialNetworkApi/Controllers/CommentsController.cs (hunk 1 above)
Comment (jhael-arce-jalafund) [reply SocialNetworkApi/Controllers/CommentsController.cs:29]: It has already been resolved

File: SocialNetworkApi/DataAccess/Repositories/Concretes/CommentRepository.cs (hunk 3 above)
Comment (jhael-arce-jalafund) [reply SocialNetworkApi/DataAccess/Repositories/Concretes/CommentRepository.cs]: It has already been resolved

File: SocialNetworkApi/DataAccess/Repositories/Concretes/CommentRepository.cs (hunk 4 above)
Comment (jhael-arce-jalafund) [reply SocialNetworkApi/DataAccess/Repositories/Concretes/CommentRepository.cs]: It has already been resolved

File: SocialNetworkApi/DataAccess/Repositories/Concretes/FriendsRepository.cs (hunk 5 above)
Comment (jhael-arce-jalafund) [reply SocialNetworkApi/DataAccess/Repositories/Concretes/FriendsRepository.cs]: It has already been resolved

File: SocialNetworkApi/Controllers/FriendsController.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
+using Microsoft.AspNetCore.Mvc;
+using SocialNetworkApi.Mappers.Request.Follow;
+using SocialNetworkApi.Services;
+
+namespace SocialNetworkApi.Controllers;
+
+[ApiController]
+[Route("api/v1/users")]
```
Comment (GabyGV-GGV) [inline SocialNetworkApi/Controllers/FriendsController.cs]: change users for friends

**Comment by GabyGV-GGV** [comment]
You have to update developed and resolve the conflict

**Comment by GabyGV-GGV** [comment]
you have several errors in importing SocialNetwork
image

**Comment by GabyGV-GGV** [comment]
Good Job!

**Review by GabyGV-GGV — APPROVED** [review]

**Pull request merged**